### ISSUE TITLE
MLX EP aborts on VibeVoice decoder CumSum shape inference

### ISSUE THREAD

**Issue by justinchuby** [justinchuby/onnxruntime-mlx#52]
## Environment

- Apple M1 Max, 32 GiB, macOS 26.6.2 arm64
- Python 3.13
- `onnxruntime==1.29.0`
- `onnxruntime-ep-mlx==0.29.5`
- VibeVoice model: `vibevoice/VibeVoice-1.5B-hf` at `edc39f80f5cae656da37baf8faa8f5502bf7081f`
- Export: Mobius staged ONNX package, float16, CPU-targeted standard ONNX attention

## Reproduction

After exporting the real-weight staged package, register the plugin and create a session for `decoder/model.onnx`:

```python
import onnxruntime as ort
import onnxruntime_ep_mlx

onnxruntime_ep_mlx.register_execution_provider_library()
session = ort.InferenceSession(
    "decoder/model.onnx",
    providers=["MLXExecutionProvider", "CPUExecutionProvider"],
)
```

The process aborts (exit code 255) before inference:

```text
MLX error: [Primitive::output_shapes] CumSum cannot infer output shapes. at mlx/c/closure.cpp:104
```

`CPUExecutionProvider` creates and runs the same decoder successfully. The decoder is required for the TTS loop, so this prevents end-to-end MLX execution even though several other VibeVoice stages register and run with MLX. A per-stage benchmark and the full error context are available from the originating test if more graph details would help.

**Comment by justinchuby** [comment]
## Exact reproducible model

This is the actual public model/export path, not a synthetic CumSum-only graph:

- Hugging Face checkpoint: [`vibevoice/VibeVoice-1.5B-hf`](https://huggingface.co/vibevoice/VibeVoice-1.5B-hf)
- Checkpoint revision: `edc39f80f5cae656da37baf8faa8f5502bf7081f`
- Exporter source: [`onnxruntime/mobius@eba3063c`](https://github.com/onnxruntime/mobius/tree/eba3063c097a5d2aa389eb503196e738d177d890)
- Export: `dtype="f16"`, `execution_provider="cpu"`, `load_weights=True`

The exported staged package contains `decoder/model.onnx` plus its external weight data. The complete public checkpoint/export is about 5.1 GiB, which exceeds a practical GitHub issue attachment; the following recreates the exact decoder locally:

```bash
git clone https://github.com/onnxruntime/mobius.git
cd mobius
git checkout eba3063c097a5d2aa389eb503196e738d177d890
pip install -e '.[testing]'
python - <<'PY'
from mobius import build

pkg = build(
    "vibevoice/VibeVoice-1.5B-hf",
    revision="edc39f80f5cae656da37baf8faa8f5502bf7081f",
    dtype="f16",
    execution_provider="cpu",
    load_weights=True,
)
pkg.save("vibevoice-onnx", max_shard_size_bytes=2_000_000_000)
PY
```

Then reproduce the plugin failure:

```python
import onnxruntime as ort
import onnxruntime_ep_mlx

onnxruntime_ep_mlx.register_execution_provider_library()
ort.InferenceSession(
    "vibevoice-onnx/decoder/model.onnx",
    providers=["MLXExecutionProvider", "CPUExecutionProvider"],
)
```

This aborts at `CumSum` shape inference on Apple Silicon with `onnxruntime-ep-mlx==0.29.5` and `onnxruntime==1.29.0`; CPU EP loads and executes this exact decoder.

**Comment by justinchuby** [comment]
## Update: CumSum is fixed; cached decode now exposes the next blocker

With `onnxruntime-ep-mlx==0.29.6.dev1`, the real VibeVoice decoder now creates an MLX session and runs its initial prefill. The complete staged pipeline also runs one audio token.

The next decoder invocation still exits the process with:

```text
MLX error: [Broadcast] Unable to infer broadcast shape at mlx/c/closure.cpp:104
```

This is a smaller reproduction than full TTS generation: use the exact public decoder export from the previous comment, create one MLX session, run one prefill with `inputs_embeds=(1,1,1536)`, `attention_mask=(1,1)`, and empty 28-layer KV caches; then run one cached decode with `inputs_embeds=(1,1,1536)`, `attention_mask=(1,2)`, `position_ids=(1,1)`, and the 28 present KV pairs from prefill. Prefill succeeds; the second `session.run()` exits with the Broadcast error.

This isolates the remaining problem to dynamic cached decode (past length 0→1 / mask length 1→2), rather than VibeVoice audio stages or the original `CumSum` session-creation failure.

**Comment by justinchuby** [comment]
## Update: dynamic cached decode `Broadcast` failure fixed

The follow-up failure was isolated to MLX's SDPA mask preparation in the shapeless compiled decode path. `ai.onnx::Attention` supplied an array mask shaped `[B,1,Q,K]`; MLX internally inserted a one-input `Broadcast` to `[B,H,Q,K]` and captured the first concrete `K` extent. Replaying the same closure after the KV cache grew then failed shape inference with:

```text
MLX error: [Broadcast] Unable to infer broadcast shape
```

The fix materializes the full SDPA mask through a two-input broadcast whose output shape is inferred from live query and mask inputs on every replay. This is restricted to shapeless token decode, so eager and shape-keyed prefill retain the compact mask and do not allocate `[B,H,Q,K]` intermediates.

Validated against the exact real-weight `vibevoice/VibeVoice-1.5B-hf` decoder: one prefill followed by cached decode now succeeds with past length `0 -> 1` and mask length `1 -> 2`. A focused regression also proves the same dynamic `Attention` partition records compile cache `MISS -> HIT` across the two calls.

Merged in #56.
